core/src/main/java/jenkins/security/UserDetailsCache.java

```diff
@@ -0,0 +1,152 @@
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package jenkins.security;
+
+import com.google.common.cache.Cache;
+import com.google.common.util.concurrent.UncheckedExecutionException;
+import hudson.Extension;
+import hudson.ExtensionList;
+import jenkins.model.Jenkins;
+import jenkins.util.SystemProperties;
+import org.acegisecurity.userdetails.UserDetails;
+import org.acegisecurity.userdetails.UsernameNotFoundException;
+import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.NoExternalUse;
+import org.springframework.dao.DataAccessException;
+
+import javax.annotation.CheckForNull;
+import javax.annotation.Nonnull;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+
+import static com.google.common.cache.CacheBuilder.newBuilder;
+
+/**
+ * Cache layer for {@link org.acegisecurity.userdetails.UserDetails} lookup.
+ *
+ * @since TODO
+ */
+@Extension
+@Restricted(NoExternalUse.class) //TODO Keep for LTS, Remove when in weekly
+public final class UserDetailsCache {
+
+    private static final String SYS_PROP_NAME = UserDetailsCache.class.getName() + ".EXPIRE_AFTER_WRITE_SEC";
+    /**
+     * Nr of seconds before a value expires after being cached, note full GC will also clear the cache.
+     * Should be able to set this value in script and then reload from disk to change in runtime.
+     */
+    private static /*not final*/ Integer EXPIRE_AFTER_WRITE_SEC = SystemProperties.getInteger(SYS_PROP_NAME, (int)TimeUnit.MINUTES.toSeconds(2));
+    private final Cache<String, UserDetails> detailsCache;
+    private final Cache<String, Boolean> existanceCache;
+
+    public UserDetailsCache() {
+        if (EXPIRE_AFTER_WRITE_SEC == null || EXPIRE_AFTER_WRITE_SEC <= 0) {
+            //just in case someone is trying to trick us
+            EXPIRE_AFTER_WRITE_SEC = SystemProperties.getInteger(SYS_PROP_NAME, (int)TimeUnit.MINUTES.toSeconds(2));
+        }
+        detailsCache = newBuilder().softValues().expireAfterWrite(EXPIRE_AFTER_WRITE_SEC, TimeUnit.SECONDS).build();
+        existanceCache = newBuilder().softValues().expireAfterWrite(EXPIRE_AFTER_WRITE_SEC, TimeUnit.SECONDS).build();
+    }
+
+    public static UserDetailsCache get() {
+        return ExtensionList.lookup(UserDetailsCache.class).get(UserDetailsCache.class);
+    }
+
+    @CheckForNull
+    public UserDetails getCached(String idOrFullName) throws UsernameNotFoundException {
```

Comment:
Well, I was actually thinking it could be good to have, as some performers might just want to do something in case the user is cached but otherwise nothing.